src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,150 @@
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "logmodel.h"
+
+#include <QApplication>
+#include <QDateTime>
+#include <QColor>
+#include <QPalette>
+
+#include "base/global.h"
+#include "base/logger.h"
+
+int BaseLogModel::rowCount(const QModelIndex &) const
+{
+    return m_messages.size();
+}
+
+int BaseLogModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    if (!index.isValid())
+        return {};
+
+    const size_t msgIndex = m_messages.size() - index.row() - 1;
+    if ((msgIndex < 0) || (msgIndex >= m_messages.size()))
+        return {};
+
+    const Message &message = m_messages[msgIndex];
+    switch (role) {
+    case Qt::ForegroundRole:
+        return message.foregroundRole;
+    case Qt::UserRole:
+        return message.userRole;
+    case TimeRole:
+        return message.time;
+    case MessageRole:
+        return message.message;
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(BaseLogModel::Message message)
+{
+    beginInsertRows(QModelIndex(), 0, 0);
+    m_messages.emplace_back(std::move(message));
+    endInsertRows();
+
+    const int count = rowCount();
+    if (count > MAX_LOG_MESSAGES) {
+        const int lastMessage = count - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        m_messages.pop_front();
+        endRemoveRows();
+    }
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_messages.clear();
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent)
+    : BaseLogModel(parent)
+{
+    for (const Log::Msg &msg : asConst(Logger::instance()->getMessages()))
+        handleNewMessage(msg);
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &log)
+{
+    Message message;
+
+    message.time = QDateTime::fromMSecsSinceEpoch(log.timestamp).toString(Qt::SystemLocaleShortDate);
+    message.message = log.message;
+
+    switch (log.type) {
+    // The RGB QColor constructor is used for performance
+    case Log::NORMAL:
+        message.foregroundRole = QApplication::palette().color(QPalette::WindowText);
+        break;
+    case Log::INFO:
+        message.foregroundRole = QColor(0, 0, 255); // blue
+        break;
+    case Log::WARNING:
+        message.foregroundRole = QColor(255, 165, 0);  // orange
+        break;
+    case Log::CRITICAL:
+        message.foregroundRole = QColor(255, 0, 0);  // red
+        break;
+    default:
+        Q_ASSERT(false);
+        break;
+    }
+
+    message.userRole = log.type;
+    addNewMessage(std::move(message));
```

Comment:
I think you should omit using move semantics to keep the code base consistent, `QVariant` uses Qt implicit sharing anyway.